mne/source_tfr.py

```diff
@@ -0,0 +1,482 @@
+        -------
+        data : The source level time-frequency transformed data.
+
+        """
+        if self._data is None:
+            # compute the solution the first time the data is accessed and
+            # remove the kernel and sensor data
+            self._remove_kernel_sens_data_()
+
+            # we can't yet give full support for TFR complex conversion
+            # if 'power' in self.method:
+            #    self._data = (self._data * self._data.conj()).real
+
+        return self._data
+
+    @data.setter
+    def data(self, value):
+        value = np.asarray(value)
+        if self._data is not None and value.ndim != self._data.ndim:
+            raise ValueError('Data array should have {} dimensions.'
+                             .format(self._data.ndim))
+
+        # vertices can be a single number, so cast to ndarray
+        if isinstance(self._vertices, list):
+            n_verts = sum([len(v) for v in self._vertices])
+        elif isinstance(self._vertices, np.ndarray):
+            n_verts = len(self._vertices)
+        else:
+            raise ValueError('Vertices must be a list or numpy array')
+
+        if value.shape[0] != n_verts:
+            raise ValueError('The first dimension of the data array must '
+                             'match the number of vertices ({0} != {1})'
+                             .format(value.shape[0], n_verts))
+
+        self._data = value
+        self._update_times()
+
+    @property
+    def shape(self):
+        """Shape of the data."""
+        if self._data is None:
+            return (self._kernel.shape[0],) + self._sens_data.shape[1:]
+
+        else:
+            return self._data.shape
+
+    @property
+    def tmin(self):
+        """The first timestamp."""
+        return self._tmin
+
+    @tmin.setter
+    def tmin(self, value):
+        self._tmin = float(value)
+        self._update_times()
+
+    @property
+    def tstep(self):
+        """The change in time between two consecutive samples (1 / sfreq)."""
+        return self._tstep
+
+    @tstep.setter
+    def tstep(self, value):
+        if value <= 0:
+            raise ValueError('.tstep must be greater than 0.')
+        self._tstep = float(value)
+        self._update_times()
+
+    @property
+    def times(self):
+        """A timestamp for each sample."""
+        return self._times
+
+    @times.setter
+    def times(self, value):
+        raise RuntimeError('You cannot write to the .times attribute directly.'
+                           ' This property automatically updates whenever '
+                           '.tmin, .tstep or .data changes.')
+
+    def _update_times(self):
+        """Update the times attribute after changing tmin, tmax, or tstep."""
+        self._times = self.tmin + (self.tstep * np.arange(self.shape[-1]))
+        self._times.flags.writeable = False
+
+    def copy(self):
+        """Return copy of SourceTFR instance."""
+        return copy.deepcopy(self)
```

Comment:
the methods / properties above should be factorized using _BaseSourceEstimate class